Add global usings for .NET projects (.NET SDK, Web and Worker)

JunTaoLuo opened the pull request:
Fixes https://github.com/dotnet/aspnetcore/issues/32451.

TODOs:
- [x] Verify that preview condition should be removed
  - [x] Ensure SDK is updated to support C# 10
- [x] Add test
- [x] Review additional usings:
  - System.ComponentModel.DataAnnotations
  - Microsoft.Extensions.Caching.Memory
  - Microsoft.AspNetCore.Mvc\
- [x] Checkin with VB expert
- [x] Perf numbers using sample minimal API Empty Web project: https://github.com/JunTaoLuo/GlobalUsings/tree/main/App

|   | Mean | Median | Std. Dev. |
| ------------- | ------------- | ------------- | ------------- |
| Before (usings in Program.cs)  | 2253.425ms | 2266.458ms | 107.8083ms |
| After (usings in generated App.ImplicitNamespaceImports.cs):  | 2221.974ms | 2214.453ms | 104.5122ms |


Followups:

- Underscore in naming: e.g. DisableImplicitNamespaceImports_Web
- Inclusion of System.Net.Http in the base SDK
- Relocation of VB Imports to props files
- Breaking change announcements
  - dotnet/Docs
  - blog post

Commit:
Add global usings for WebSdk projects

JunTaoLuo commented on src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.targets:
```diff
@@ -84,4 +95,17 @@ Copyright (c) .NET Foundation. All rights reserved.
     </When>
   </Choose>
 
+  <Target Name="_CopyDefaultGlobalUsingFiles" AfterTargets="Restore" Condition="'$(EnableGlobalUsingFiles)' == 'true'">
```
Compute outputs to better support incremental builds.

JunTaoLuo commented on src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.targets:
```diff
@@ -22,8 +22,19 @@ Copyright (c) .NET Foundation. All rights reserved.
+      And '$(TargetFrameworkIdentifier)' == '.NETCoreApp'
+      And '$(_TargetFrameworkVersionWithoutV)' != ''
+      And '$(_TargetFrameworkVersionWithoutV)' &gt;= '6.0'
+      And '$(LangVersion)' == 'preview'">true</EnableGlobalUsingFiles>
```
Remove the preview requirement when C# 10 is default.

JunTaoLuo commented on src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.targets:
```diff
@@ -22,8 +22,19 @@ Copyright (c) .NET Foundation. All rights reserved.
+      '$(EnableGlobalUsingFiles)' == ''
+      And '$(TargetFrameworkIdentifier)' == '.NETCoreApp'
+      And '$(_TargetFrameworkVersionWithoutV)' != ''
+      And '$(_TargetFrameworkVersionWithoutV)' &gt;= '6.0'
```
Should we add any more restrictions? For example, should we restrict global usings to apps only and do not add these for libraries?

JunTaoLuo commented on src/WebSdk/ProjectSystem/content/Microsoft.NET.Sdk.Web.ProjectSystem.DefaultGlobalUsing.cs:
```diff
@@ -0,0 +1,15 @@
+global using System;
```
This list is based on https://github.com/dotnet/aspnetcore/issues/32451#issue-877008542. Though I've removed EF from the list.

JunTaoLuo commented on src/WebSdk/ProjectSystem/content/Microsoft.NET.Sdk.Web.ProjectSystem.DefaultGlobalUsing.cs:
```diff
@@ -0,0 +1,15 @@
+global using Microsoft.Extensions.Configuration;
+global using Microsoft.Extensions.DependencyInjection;
+global using Microsoft.Extensions.Hosting;
+global using Microsoft.Extensions.Logging;
```
Include license header once design/list is approved.

Commit:
add global::

davidfowl commented on src/WebSdk/ProjectSystem/content/Microsoft.NET.Sdk.Web.ProjectSystem.DefaultGlobalUsing.cs:
```diff
@@ -0,0 +1,15 @@
+global using global::System;
```
Not this not the assembly info file

pranavkm commented on src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.targets:
```diff
@@ -84,4 +95,17 @@ Copyright (c) .NET Foundation. All rights reserved.
+
+  <Target Name="_AddDefaultGlobalUsingFilesToCompile" BeforeTargets="BeforeCompile" DependsOnTargets="_CopyDefaultGlobalUsingFiles" Condition="'$(EnableGlobalUsingFiles)' == 'true'">
+    <ItemGroup>
+      <Compile Include="$(GlobalUsingFilesDirectory)*.cs" />
```
Btw, if we're dropping the file copying stuff, you can Compile Include it from the location in the SDK. 
```suggestion
      <Compile Include="$(MSBuildThisFileDirectory)Microsoft.NET.Sdk.Web.DefaultGlobalUsing.cs" />
```

Commit:
Include global using as compile item group

Commit:
Update

JunTaoLuo commented on src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.DefaultGlobalUsing.cs:
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) .NET Foundation. All rights reserved.
```
Two questions:

1. Does it make sense for us to include a license header in files we intend to inject into user's projects? I'm not sure about licensing implications
2. If we should, is this the correct license header?

pranavkm replied:
I don't think we should. The generated assembly info files do not, so this should be similar.

JunTaoLuo replied:
Do we need an auto-generated header? Or should we not have any headers at all?

halter73 replied:
If there is a header, it should probably tell you how to disable this. Not sure who would actually see this though. You'd have to be looking through the obj folder, right?

JunTaoLuo commented on src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.DefaultGlobalUsing.cs:
```diff
@@ -0,0 +1,18 @@
+global using global::Microsoft.Extensions.Configuration;
+global using global::Microsoft.Extensions.DependencyInjection;
+global using global::Microsoft.Extensions.Hosting;
+global using global::Microsoft.Extensions.Logging;
```
@vijayrkn I checked that Microsoft.NET.Sdk.Web.ProjectSystem.*.nupkg contains this file. I assume I don't have to do anything else to ensure *.cs files are included in the appropriate places?

vijayrkn replied:
Once it is included in the sdk, then you should be able to use it without any other changes.

JunTaoLuo commented on src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.targets:
```diff
@@ -22,8 +22,17 @@ Copyright (c) .NET Foundation. All rights reserved.
     <MSBuildWebTargetsPath>$(MSBuildExtensionsPath)\Microsoft\VisualStudio\Managed.Web\</MSBuildWebTargetsPath>
     <AspNetCoreHostingModel Condition="'$(AspNetCoreHostingModel)' == '' And '$(TargetFrameworkIdentifier)' == '.NETCoreApp' And '$(_TargetFrameworkVersionWithoutV)' != '' And '$(_TargetFrameworkVersionWithoutV)' &gt;= '3.0' ">inprocess</AspNetCoreHostingModel>
     <AspNetCoreModuleName Condition="'$(AspNetCoreModuleName)' == '' And '$(TargetFrameworkIdentifier)' == '.NETCoreApp' And '$(_TargetFrameworkVersionWithoutV)' != '' And '$(_TargetFrameworkVersionWithoutV)' &gt;= '3.0' ">AspNetCoreModuleV2</AspNetCoreModuleName>
+    <EnableAspNetCoreGlobalUsings Condition="
```
@vijayrkn I don't see any tests for the WebSDK. I've been doing spot checks with https://github.com/JunTaoLuo/GlobalUsings. Is that sufficient to verify this change?

vijayrkn replied:
There are unit tests for the tasks & end-end tests for the folder publish. Let me check the status of those.

vijayrkn replied:
@JunTaoLuo  - I think we need to add something similar to this https://github.com/dotnet/sdk/blob/main/src/Tests/Microsoft.NET.Sdk.Publish.Tasks.Tests/EndToEnd/FolderPublish50.cs but rather than publish, it should cover 'dotnet run' for web.

JunTaoLuo replied:
Tests added

Commit:
Add Ifdef

JunTaoLuo commented on src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.DefaultGlobalUsing.cs:
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#if NET6_0
```
I suppose we should start the ifdef pattern here even though we don't technically need it in the first release. I assume we want an error to remind ourselves to update this file.

davidfowl replied:
What is the error for?

JunTaoLuo replied:
Remind ourselves to update the ifdefs so each TFM has a new section. I'm hoping somewhere in our test pipeline this error will show up when we update the TFM.

pranavkm replied:
do you want to say net60 or greater? It's fairly unlikely this list is going to change very often, no?

davidfowl replied:
Exactly

JunTaoLuo replied:
FYI, the conditionality of these usings are now specified in the target files with conditions like `'$(TargetFrameworkIdentifier)' == '.NETCoreApp'
                      and $([MSBuild]::VersionGreaterThanOrEquals($(_TargetFrameworkVersionWithoutV), '6.0'))`

davidfowl commented on src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.targets:
```diff
@@ -22,8 +22,17 @@ Copyright (c) .NET Foundation. All rights reserved.
     <AspNetCoreHostingModel Condition="'$(AspNetCoreHostingModel)' == '' And '$(TargetFrameworkIdentifier)' == '.NETCoreApp' And '$(_TargetFrameworkVersionWithoutV)' != '' And '$(_TargetFrameworkVersionWithoutV)' &gt;= '3.0' ">inprocess</AspNetCoreHostingModel>
     <AspNetCoreModuleName Condition="'$(AspNetCoreModuleName)' == '' And '$(TargetFrameworkIdentifier)' == '.NETCoreApp' And '$(_TargetFrameworkVersionWithoutV)' != '' And '$(_TargetFrameworkVersionWithoutV)' &gt;= '3.0' ">AspNetCoreModuleV2</AspNetCoreModuleName>
+    <EnableAspNetCoreGlobalUsings Condition="
+      '$(EnableGlobalUsingFiles)' == ''
```
```suggestion
      '$(EnableAspNetCoreGlobalUsings)' == ''
```
?

Commit:
Update src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.targets

Co-authored-by: David Fowler <davidfowl@gmail.com>

davidfowl commented on src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.DefaultGlobalUsing.cs:
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#if NET6_0
```
```suggestion
#if NET6_0_OR_GREATER
```

JunTaoLuo replied:
🤦

JunTaoLuo replied:
Oops, wrong comment, the facepalm is for the other typo fix.

JunTaoLuo replied:
Super cool! I haven't seen this feature used anywhere yet: https://github.com/dotnet/designs/blob/main/accepted/2020/or-greater-defines/or-greater-defines.md.

davidfowl commented on src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.DefaultGlobalUsing.cs:
```diff
@@ -0,0 +1,22 @@
+global using global::Microsoft.Extensions.Hosting;
+global using global::Microsoft.Extensions.Logging;
+#else
+#error Target frameworks need to be updated.
```
@JunTaoLuo does this file show up in the solution explorer? Do we want it to?

cc @DamianEdwards

JunTaoLuo replied:
That's a good point. I'll see what I can do here. Theoretically, we don't want the users to edit this at all but I don't know how we'd make a cs file uneditable. Alternatively, we can write a MSBuild task that generates this file and include it as a dll. That would make this file uneditable and nothing would show up in the solution explorer.

davidfowl replied:
Yes, look at what is being done today for assembly info, and do that. Try this experience though.

JunTaoLuo replied:
Those are generated by code, I think. I saw the auto-generated headers but I can confirm tomorrow.

JunTaoLuo replied:
File removed. It's now essentially the same logic as AssemblyInfo.cs. It doesn't show up in the solution explorer and is not editable.

davidfowl commented:
2 more considerations for global usings:
- System.ComponentModel.DataAnnotations
- Microsoft.Extensions.Caching.Memory

JunTaoLuo commented:
>System.ComponentModel.DataAnnotations
Microsoft.Extensions.Caching.Memory

Hmm, what are the common use cases for these two? I'd assume MVC stuff for S.CM.DA but for M.E.C.M is it really used that frequently?

JunTaoLuo commented:
Ah I found some justification for the two new usings in offline conversations, colour me somewhat convinced.

vijayrkn commented on src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.DefaultGlobalUsing.cs:
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) .NET Foundation. All rights reserved.
```
@JunTaoLuo - Can we move this out of the Targets folder into a separate folder called "DefaultContent" or something similar. This folder is meant for all msbuild targets.

JunTaoLuo replied:
Sure, that's actually what I had originally, but I figured I'll keep everything together until someone tells me otherwise 😄

vijayrkn replied:
wish granted :)

JunTaoLuo replied:
File removed.

halter73 commented on src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.targets:
```diff
@@ -22,8 +22,17 @@ Copyright (c) .NET Foundation. All rights reserved.
+      '$(EnableAspNetCoreGlobalUsings)' == ''
+      And '$(TargetFrameworkIdentifier)' == '.NETCoreApp'
+      And $([MSBuild]::VersionGreaterThanOrEquals('$(TargetFrameworkVersion)', '6.0'))
+      And '$(LangVersion)' == 'preview'">true</EnableAspNetCoreGlobalUsings>
```
@shirhatti asked me if this will cause a problem for F#, and I'm not sure. Do we add a condition to skip this for fsproj's targeting "Microsoft.NET.Sdk.Web"?

I tried to `<Compile Include=` a DefaultGlobalUsing.cs in an fsproj just to see what happens.

```
FSC : error FS0226: The file extension of 'Microsoft.NET.Sdk.Web.ProjectSystem.DefaultGlobalUsing.cs' is not recognized. Source files must have extension .fs, .fsi, .fsx, .fsscript, .ml or .mli. [C:\dev\halter73\HoudiniPlaygroundFSharp\HoudiniPlaygroundFSharp.fsproj]
```

I did not test with the actual SDK changes, so it's possible I missing some reason this shouldn't affectt fsproj's.

DamianEdwards replied:
The LangVersion check won't pass eventually right, as the C# 10 just becomes the default and is not preview anymore, i.e. when we GA. Also to @halter73's point VB has its own concept of global imports as this is language feature, not general SDK feature, so the condition should ensure it's only applying to C# projects.

davidfowl replied:
This is going into the real sdk now anyways.

JunTaoLuo replied:
Added the condition for C# only. Still need to check whether we can remove the preview condition though.

Commit:
Refactor <App>.GlobalUsings.cs generation to M.NET.Sdk

JunTaoLuo commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateGlobalUsings.targets:
```diff
@@ -0,0 +1,64 @@
+    -->
+  <PropertyGroup>
+    <GeneratedGlobalUsingsFile Condition="'$(GeneratedGlobalUsingsFile)' ==''">$(IntermediateOutputPath)$(MSBuildProjectName).GlobalUsings$(DefaultLanguageSourceExtension)</GeneratedGlobalUsingsFile>
+    <GenerateGlobalUsings Condition="'$(GenerateGlobalUsings)' == '' and '$(LangVersion)' == 'preview'">true</GenerateGlobalUsings>
```
Currently the logic defaults to true when:
1. Is a C# project (condition on import in Microsoft.NET.Sdk.targets)
2. LangVersion is set to preview (we can probably remove this in preview7, I just need to test with the latest SDK)
3. It's not explicitly disabled by the user
4. The TFM is net6.0 or higher (condition on the ItemGroup in this file and in Microsoft.NET.Sdk.Web.ProjectSystem.targets)

We should consider the alternative where this is an opt in feature and our `web api` and `empty web` templates sets `GenerateGlobalUsings` to true.

DamianEdwards replied:
The plan is for this to be enabled by default.

JunTaoLuo commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateGlobalUsings.targets:
```diff
@@ -0,0 +1,64 @@
+
+  <ItemGroup Condition=" '$(TargetFrameworkIdentifier)' == '.NETCoreApp'
+                      and $([MSBuild]::VersionGreaterThanOrEquals($(_TargetFrameworkVersionWithoutV), '6.0'))">
+    <GlobalUsings Include="System" />
```
Generated file for projects targeting `Microsoft.NET.Sdk` e.g. [GlobalUsing/Library](https://github.com/JunTaoLuo/GlobalUsings/tree/main/Library)
```c#
// <autogenerated />
global using global::System;
global using global::System.Collections.Generic;
global using global::System.Linq;
global using global::System.Net.Http;
global using global::System.Net.Http.Json;
```

Generated file for projects targeting `Microsoft.NET.Sdk.Web` e.g. [GlobalUsing/App](https://github.com/JunTaoLuo/GlobalUsings/tree/main/App)
```c#
// <autogenerated />
global using global::System;
global using global::System.Collections.Generic;
global using global::System.Linq;
global using global::System.Net.Http;
global using global::System.Net.Http.Json;
global using global::Microsoft.AspNetCore.Builder;
global using global::Microsoft.AspNetCore.Hosting;
global using global::Microsoft.AspNetCore.Http;
global using global::Microsoft.AspNetCore.Mvc;
global using global::Microsoft.AspNetCore.Routing;
global using global::Microsoft.Extensions.Configuration;
global using global::Microsoft.Extensions.DependencyInjection;
global using global::Microsoft.Extensions.Hosting;
global using global::Microsoft.Extensions.Logging;
```

pranavkm replied:
FYI @KathleenDollard / @terrajobst since this affects every SDK type project.

Commit:
Fix spacing

nil4 commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateGlobalUsings.targets:
```diff
@@ -0,0 +1,64 @@
+      <_GlobalUsingsLines Include="global using global::%(GlobalUsings.Identity)%3B"/>
+    </ItemGroup>
+
+    <WriteLinesToFile Lines="@(_GlobalUsingsLines)" File="$(GeneratedGlobalUsingsFile)" Overwrite="true" />
```
I may be misunderstanding, but this appears to (over)write the file every time. Is that not likely to cause builds to no longer be incremental, since the global usings file is always newer than earlier build outputs?

JunTaoLuo commented on src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.targets:
```diff
@@ -35,6 +35,18 @@ Copyright (c) .NET Foundation. All rights reserved.
     <ProjectCapability Include="AspNetCoreInProcessHosting" />
   </ItemGroup>
 
+  <ItemGroup Condition=" '$(TargetFrameworkIdentifier)' == '.NETCoreApp' And '$(_TargetFrameworkVersionWithoutV)' != '' And '$(_TargetFrameworkVersionWithoutV)' &gt;= '6.0' ">
```
TargetFrameworkVersion is v6.0. I think this works with _TargetFrameworkVersionWithoutV though

JunTaoLuo commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateGlobalUsings.targets:
```diff
@@ -0,0 +1,64 @@
+    Targets that generate Compile items are also expected to run before
+    BeforeCompile targets (common targets convention).
+   -->
+  <Target Name="GenerateGlobalUsings"
```
I've been following what GenerateAssemblyInfo does. I think this is not private because the target may be used in Before/AfterTargets and DependsOn since end users may need to modify the GlobalUsing itemgroup and need to control the order.

Commit:
Feedback

JunTaoLuo commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:
```diff
@@ -1073,6 +1073,7 @@ Copyright (c) .NET Foundation. All rights reserved.
   <Import Project="$(MSBuildThisFileDirectory)Microsoft.NET.DisableStandardFrameworkResolution.targets" Condition="'$(DisableStandardFrameworkResolution)' == 'true'" />
   <Import Project="$(MSBuildThisFileDirectory)Microsoft.NET.DesignerSupport.targets" />
   <Import Project="$(MSBuildThisFileDirectory)Microsoft.NET.GenerateAssemblyInfo.targets" Condition="'$(UsingNETSdkDefaults)' == 'true'"/>
+  <Import Project="$(MSBuildThisFileDirectory)Microsoft.NET.GenerateGlobalUsings.targets" Condition="'$(Language)' == 'C#'"/>
```
Hmm how about we use a more global setting to control whether these targets are imported like `EnableGlobalUsings`.

And for .NET SDK and WebSDK we have more fine grained properties like `GenerateDotNetGlobalUsings` and `GenerateAspNetCoreGlobalUsings`?

DamianEdwards replied:
Agreed on the pattern, but I think the naming needs work as `EnableGlobalUsings` sounds like the compiler feature itself is being enabled/disabled, and the others should reflect the SDK they're in (e.g. web vs. ASP.NET Core).

Perhaps the following instead:
- `EnableDefaultGlobalUsings` enables/disables all default global usings (can be overridden by below properties)
- `EnableDefaultGlobalUsings_DotNet` enables/disables default global usings defined in **Microsoft.NET.Sdk**
- `EnableDefaultGlobalUsings_Web` enables/disables default global usings defined in **Microsoft.NET.Sdk.Web**
- `EnableDefaultGlobalUsings_Razor` enables/disables default global usings defined in **Microsoft.NET.Sdk.Razor**
- `EnableDeafultGlobalUsings_WindowsDesktop` enables/disables default global usings defined in **Microsoft.NET.Sdk.WindowsDesktop**
- `EnableDeafultGlobalUsings_Worker` enables/disables default global usings defined in **Microsoft.NET.Sdk.Worker**
- `EnableDeafultGlobalUsings_BlazorWebAssembly` enables/disables default global usings defined in **Microsoft.NET.Sdk.BlazorWebAssembly**

JunTaoLuo replied:
I like it, I'll update the Dotnet/Web/Worker ones. I think once we have the pattern, we can get experts in each area to figure out the list of namespaces to include.

JunTaoLuo replied:
@DamianEdwards I thought about this a little more and I don't know if the override we discussed is the best approach. Specifically I find the scenario where the top level option `EnableDefaultGlobalUsings` is `false` and `EnableDefaultGlobalUsings_Web` is `true` generating a file with the AspNetCore global usings make sense.

If the goal is to disable the .NET global usings but retain the AspNetCore global usings, it makes sense to keep `EnableDefaultGlobalUsings` and `EnableDefaultGlobalUsings_Web` as `true` and explicitly set `EnableDefaultGlobalUsings_DotNet` to `false`. In this model, the top level `EnableDefaultGlobalUsings` controls whether the entire feature is turned on or off while the SDK specific property controls whether default global usings specific to that SDK are included when generating the file.

To cite precedence. This is the model that's used to control how `GenerateAssemblyInfo` works. The top level `GenerateAssemblyInfo` controls whether the file is generated at all: https://github.com/dotnet/sdk/blob/main/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateAssemblyInfo.targets#L22. Then individual properties control whether each attribute is included when generating the file: https://github.com/dotnet/sdk/blob/main/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateAssemblyInfo.targets#L26-L40.

JunTaoLuo replied:
If you agree, I'll update the logic here so that we don't add any namespaces to the itemgroup at all when the top level flag is set to false.

JunTaoLuo replied:
Spoke offline and we'd like to adopt the logic I described above.

Commit:
Add global usings for workers

JunTaoLuo commented:
FYI @KathleenDollard @marcpopMSFT @wli3 I'd like to chat about this feature at the meeting tomorrow. I'll go into details when we meet but feel free to look this over before then.

JunTaoLuo commented:
Review:

- No to M.A.Mvc 
- No to System.ComponentModel.DataAnnotations
- No to M.E.Caching.Memory

Commit:
Remove preview condition and remove M.A.Mvc

Commit:
Renames based on recommendations

Commit:
Fix typo and make top level option controll enabling of entire feature

JunTaoLuo commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateGlobalUsings.targets:
```diff
@@ -0,0 +1,65 @@
+  <ItemGroup Condition=" '$(EnableDefaultGlobalUsings_DotNet)' == 'true'
+                      and '$(TargetFrameworkIdentifier)' == '.NETCoreApp'
+                      and $([MSBuild]::VersionGreaterThanOrEquals($(_TargetFrameworkVersionWithoutV), '6.0'))">
+    <DefaultGlobalUsing Include="System" />
```
Add System.IO (commonly used).
Add System.Threading.Tasks (need to be used for System.Net.Http, used in Framework template).
Not yet for System.Text.Json (System.Net.Http scenarios)
Not yet for System.Text.Json.Nodes (System.Net.Http scenarios)
Not yet for System.Text.Json.Serialization (System.Net.Http scenarios)
Remove System.Net.Http.Json

JunTaoLuo commented:
FYI test failures are due to global usings still being recognized as a preview feature:

```
C:\h\w\B3F8095C\t\dotnetSdkTests\curzqgv4.sjw\Microsoft.Asp---72D13848\WebTest\obj\Debug\netcoreapp2.2\win10-x64\WebTest.DefaultGlobalUsings.cs(2,1): error CS8652: The feature 'global using directive' is currently in Preview and unsupported. To use Preview features, use the 'preview' language version. 
```

I'm tracking down when this will be updated here after Roslyn changes were made last week in https://github.com/dotnet/roslyn/pull/54359

Commit:
Update namespaces added by .NET SDK

Commit:
Renames to align with VB

Commit:
More renaming for consistency

RussKie commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateImplicitNamespaceImports.targets:
```diff
@@ -0,0 +1,66 @@
+    <Import Include="System.Collections.Generic" />
+    <Import Include="System.IO" />
+    <Import Include="System.Linq" />
+    <Import Include="System.Net.Http" />
```
There's a long email thread, which you may have missed.
tl;dr: `Import` is an established way of defining global imports in Visual Basic, https://github.com/dotnet/sdk/blob/main/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.VisualBasic.targets#L16-L39, and there's no reason to create a new one.

wli3 replied:
looks like part of the feature. If PMs signed off that should be good. Any thoughts or bug would happen if there are duplication or someone put this in directoy.build.prop?

JunTaoLuo replied:
We allow the user to modify this list in their directory.build.props. It's a publicly available extension point.

wli3 commented:
@dsplaisted do you have the context and the time? I realize I will take some time to read the whole thread.

RussKie commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateImplicitNamespaceImports.targets:
```diff
@@ -0,0 +1,66 @@
+          Outputs="$(GeneratedImplicitNamespaceImportFile)">
+    <ItemGroup>
+      <_ImportFileLine Include="// %3Cautogenerated />"/>
+      <_ImportFileLine Include="global using global::%(Import.Identity)%3B"/>
```
How would we handle duplicates? I'd go with a distinct option. Otherwise we need to have a meaningful error message.

JunTaoLuo replied:
I can add a warning for duplicated names and use a distinct list.

RussKie commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateImplicitNamespaceImports.targets:
```diff
@@ -0,0 +1,66 @@
+    <Import Include="System.Collections.Generic" />
+    <Import Include="System.IO" />
+    <Import Include="System.Linq" />
+    <Import Include="System.Net.Http" />
```
I don't think this should be defined globally. Console or Windows desktop apps don't always use it.

JunTaoLuo replied:
When we reviewed the list, we felt that System.Net.Http, while not used in all apps, are used frequently enough that we want to include it by default. cc @halter73 @terrajobst

RussKie commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateImplicitNamespaceImports.targets:
```diff
@@ -0,0 +1,66 @@
+    Targets that generate Compile items are also expected to run before
+    BeforeCompile targets (common targets convention).
+   -->
+  <Target Name="GenerateImplicitNamespaceImports"
+          BeforeTargets="BeforeCompile;CoreCompile"
+          DependsOnTargets="PrepareForBuild;CoreGenerateImplicitNamespaceImports"
+          Condition="'$(DisableImplicitNamespaceImports)' != 'true'" />
+
+  <Target Name="CoreGenerateImplicitNamespaceImports"
+          Outputs="$(GeneratedImplicitNamespaceImportFile)">
+    <ItemGroup>
+      <_ImportFileLine Include="// %3Cautogenerated />"/>
+      <_ImportFileLine Include="global using global::%(Import.Identity)%3B"/>
+    </ItemGroup>
+
+    <WriteLinesToFile Lines="@(_ImportFileLine)" File="$(GeneratedImplicitNamespaceImportFile)" Overwrite="true" WriteOnlyWhenDifferent="true" />
+
+    <ItemGroup>
+      <Compile Include="$(GeneratedImplicitNamespaceImportFile)" />
+      <FileWrites Include="$(GeneratedImplicitNamespaceImportFile)" />
+    </ItemGroup>
+  </Target>
```
I applied this locally, and something is not quite right here - it applies to non net6.0 builds as well, e.g. I tried building a netcoreapp3.1
![image](https://user-images.githubusercontent.com/4403806/124074688-359f3b00-da87-11eb-9dc4-f1dcb2e85d82.png)

JunTaoLuo replied:
Ah I see, I should add a condition to only write when the itemgroup is not empty.

JunTaoLuo replied:
The logic is updated to handle when the itemgroup is empty. A test case was also added.

RussKie commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateImplicitNamespaceImports.targets:
```diff
@@ -0,0 +1,66 @@
+    -->
+  <PropertyGroup>
+    <GeneratedImplicitNamespaceImportFile Condition="'$(GeneratedImplicitNamespaceImportFile)' ==''">$(IntermediateOutputPath)$(MSBuildProjectName).ImplicitNamespaceImports$(DefaultLanguageSourceExtension)</GeneratedImplicitNamespaceImportFile>
+    <DisableImplicitNamespaceImports_DotNet Condition="'$(DisableImplicitNamespaceImports_DotNet)' == '' and '$(DisableImplicitNamespaceImports)' != 'true'">false</DisableImplicitNamespaceImports_DotNet>
+  </PropertyGroup>
+
+  <ItemGroup Condition=" '$(DisableImplicitNamespaceImports_DotNet)' != 'true'
+                      and '$(TargetFrameworkIdentifier)' == '.NETCoreApp'
+                      and $([MSBuild]::VersionGreaterThanOrEquals($(_TargetFrameworkVersionWithoutV), '6.0'))">
+    <Import Include="System" />
+    <Import Include="System.Collections.Generic" />
+    <Import Include="System.IO" />
+    <Import Include="System.Linq" />
+    <Import Include="System.Net.Http" />
+    <Import Include="System.Threading.Tasks" />
+  </ItemGroup>
+
+  <!--
+    Note that this must run before every invocation of CoreCompile to ensure that all compiler
+    runs see the generated assembly info. There is at least one scenario involving Xaml
+    where CoreCompile is invoked without other potential hooks such as Compile or CoreBuild,
+    etc., so we hook directly on to CoreCompile. Furthermore, we  must run *after*
+    PrepareForBuild to ensure that the intermediate directory has been created.
+
+    Targets that generate Compile items are also expected to run before
+    BeforeCompile targets (common targets convention).
+   -->
+  <Target Name="GenerateImplicitNamespaceImports"
+          BeforeTargets="BeforeCompile;CoreCompile"
+          DependsOnTargets="PrepareForBuild;CoreGenerateImplicitNamespaceImports"
+          Condition="'$(DisableImplicitNamespaceImports)' != 'true'" />
```
This makes a multi target app build correctly, and emit only for net6.0:
```suggestion
    <DisableImplicitNamespaceImports_DotNet Condition="'$(DisableImplicitNamespaceImports_DotNet)' == '' and '$(DisableImplicitNamespaceImports)' != 'true'">false</DisableImplicitNamespaceImports_DotNet>
    <EnableGlobalUsings>false</EnableGlobalUsings>
    <EnableGlobalUsings Condition=" '$(DisableImplicitNamespaceImports_DotNet)' != 'true'
                                    and '$(TargetFrameworkIdentifier)' == '.NETCoreApp'
                                    and $([MSBuild]::VersionGreaterThanOrEquals($(_TargetFrameworkVersionWithoutV), '6.0'))">true</EnableGlobalUsings>
  </PropertyGroup>

  <ItemGroup Condition=" '$(EnableGlobalUsings)' == 'true'">
    <Import Include="System" />
    <Import Include="System.Collections.Generic" />
    <Import Include="System.IO" />
    <Import Include="System.Linq" />
    <Import Include="System.Net.Http" />
    <Import Include="System.Threading.Tasks" />
  </ItemGroup>

  <!--
    Note that this must run before every invocation of CoreCompile to ensure that all compiler
    runs see the generated assembly info. There is at least one scenario involving Xaml
    where CoreCompile is invoked without other potential hooks such as Compile or CoreBuild,
    etc., so we hook directly on to CoreCompile. Furthermore, we  must run *after*
    PrepareForBuild to ensure that the intermediate directory has been created.

    Targets that generate Compile items are also expected to run before
    BeforeCompile targets (common targets convention).
   -->
  <Target Name="GenerateImplicitNamespaceImports"
          BeforeTargets="BeforeCompile;CoreCompile"
          DependsOnTargets="PrepareForBuild;CoreGenerateImplicitNamespaceImports"
          Condition=" '$(EnableGlobalUsings)' == 'true'" />
```

JunTaoLuo replied:
I think I prefer the itemgroup not empty approach.

RussKie replied:
Apologies, I'm not sure what you mean.

RussKie commented on src/WebSdk/Worker/Targets/Microsoft.NET.Sdk.Worker.targets:
```diff
@@ -25,4 +26,14 @@ Copyright (c) .NET Foundation. All rights reserved.
   </ItemGroup>
 
+
+  <ItemGroup Condition=" '$(DisableImplicitNamespaceImports_Worker)' != 'true'
+      And '$(TargetFrameworkIdentifier)' == '.NETCoreApp'
+      And $([MSBuild]::VersionGreaterThanOrEquals('$(_TargetFrameworkVersionWithoutV)', '6.0'))">
```
I don't believe this condition is required anywhere besides Microsoft.NET.GenerateImplicitNamespaceImports.targets (and probably only needed on `GenerateImplicitNamespaceImports` target). 
Few reasons:
- nothing will be emitted unless the target is run, and
- this makes it extremely brittle - if the condition changes it will need to be applied universally and in multiple places, including SDKs that aren't part of this repo (e.g. Windows Desktop, which resides in dotnet/wpf and dotnet/windowsdesktop, but has bits ingested from dotnet/winforms).

RussKie replied:
E.g. this is how I expect the Windows Forms global usings props to look:
```xml
<Project>

  <ItemGroup Condition=" '$(UseWindowsForms)' == 'true' ">
    <!--
      SDK defines the following global usings:
      * System
      * System.Collections.Generic
      * System.Linq
      * System.Threading.Tasks
    -->
    <Import Include="System.Drawing" />
    <Import Include="System.Windows.Forms" />
  </ItemGroup>

</Project>
```

Referenced through the Microsoft.NET.Sdk.WindowsDesktop.props this leads to the following:
![image](https://user-images.githubusercontent.com/4403806/124079995-2c18d180-da8d-11eb-82aa-e7d9184dd1f9.png)

JunTaoLuo replied:
We actually want each SDK to specify their own list of namespaces that are TFM driven to ensure compatibility. Any changes to the list of namespaces are breaking changes so we want to guard them with TFM checks in each SDK.

I think it's valuable to add a TFM check for > net6.0 for the entire feature though. 

FYI the CS0185 warning is going to be removed. Let me find the issue and link it here.

RussKie replied:
> We actually want each SDK to specify their own list of namespaces that are TFM driven to ensure compatibility. Any changes to the list of namespaces are breaking changes so we want to guard them with TFM checks in each SDK.

Alright. Though IMO it should be additive to the base list.

dsplaisted replied:
I think each SDK that adds `Import` items should condition it on the target framework identifier and version.  That allows us to change the default imports over time without breaking existing projects (until they are retargeted).

I think these `Import` items should be moved to `Microsoft.NET.Sdk.Worker.props`.

RussKie reviewed (dismissed):
Several issues that must be addressed

RussKie commented:
We also need to address this:
```
C:\Development\!Tests\ProjectConfiguration\obj\Debug\net6.0-windows\ProjectConfiguration.AssemblyInfo.cs(10,7): warning CS0105: The using directive for 'System' appeared previously in this namespace 
```

Commit:
Unblock test writing

Commit:
Try rebasing

Commit:
Update conditions

Commit:
Add unit tests

JunTaoLuo commented on src/Assets/TestProjects/TestConsoleApp/ConsoleApp.csproj:
```diff
@@ -0,0 +1,10 @@
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net6.0</TargetFramework>
+    <LangVersion>preview</LangVersion>
```
Ignore this for now, I'm still figuring out why Roslyn changes for C# 10 still isn't working properly in this repo. This will be removed before this change is merged.

davidfowl replied:
https://github.com/dotnet/aspnetcore/pull/33393/files

Look for something pinning Microsoft.Net.Compilers.Toolset e.g. https://github.com/dotnet/sdk/blob/9f265a2db3d9c9bcd9c1ec0eeb7fc25f1a8f001b/eng/Versions.props#L120

Commit:
Add license headers

JunTaoLuo commented on src/WebSdk/Worker/Targets/Microsoft.NET.Sdk.Worker.targets:
```diff
@@ -25,4 +26,15 @@ Copyright (c) .NET Foundation. All rights reserved.
+  <ItemGroup Condition=" '$(DisableImplicitNamespaceImports_Worker)' != 'true'
+      And '$(TargetFrameworkIdentifier)' == '.NETCoreApp'
+      And $([MSBuild]::VersionGreaterThanOrEquals('$(_TargetFrameworkVersionWithoutV)', '6.0'))">
+    <Import Include="System.Threading" />
```
Should this be added to the base .NET SDK instead of here?

RussKie replied:
+1

Commit:
Add test for duplicated namespaces

dsplaisted commented on src/Assets/TestProjects/TestConsoleApp/ConsoleApp.csproj:
```diff
@@ -0,0 +1,10 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
For simple projects like this, I prefer the `TestProject` APIs rather than adding new assets under the TestProjects folder.  You can set project properties and source files in code.

Can we switch to doing that instead of adding more TestProjects assets?

JunTaoLuo replied:
These are a bit different in that the *.cs files don't have any using statements. Though it's possible to modify *.cs files in code it's not well supported by `TestAsset` I think.

dsplaisted replied:
You can use the `SourceFiles` property of `TestProject`:

```c#
            testProject.SourceFiles["Program.cs"] = @"
//  Top level statement
Console.WriteLine(""Hello World!"");
```

JunTaoLuo replied:
I'll update the tests after I figure out the language version issues.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateImplicitNamespaceImports.targets:
```diff
@@ -0,0 +1,68 @@
+    <DisableImplicitNamespaceImports_DotNet Condition="'$(DisableImplicitNamespaceImports_DotNet)' == '' and '$(DisableImplicitNamespaceImports)' != 'true'">false</DisableImplicitNamespaceImports_DotNet>
+  </PropertyGroup>
+
+  <ItemGroup Condition=" '$(DisableImplicitNamespaceImports_DotNet)' != 'true'
```
Can we put this `ItemGroup` in a `.props` file instead of a `.targets` file?  That way it's simpler to remove an `Import` you don't want in the body of a project file:

```xml
<ItemGroup>
    <Import Remove="System.Threading.Tasks"/>
</ItemGroup>
```

This works because of MSBuilds multi-pass evaluation.  Properties are evaluated in the first pass, so you can put items in `.props` files that have conditions based on properties that are set "later".

I think we should also move the [VB import definitions](https://github.com/dotnet/sdk/blob/cfc3463578c94dba51115d8fc218a4a7d96281c2/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.VisualBasic.targets#L16-L39) from Microsoft.NET.Sdk.VisualBasic.targets to the same .props file.

KathleenDollard replied:
I do not think we should move the VB Import definitions this late in .NET 6, unless it is super evident that it cannot break anything.

JunTaoLuo replied:
>  Properties are evaluated in the first pass, so you can put items in .props files that have conditions based on properties that are set "later".

@dsplaisted I think I'm missing something here, how does this work? I originally placed the Import items in the .target files since they needed to be TFM driven. Seems like you are suggesting that it's possible to do that even if the itemgroup is defined in .props files and I'd like to make sure I understand how that works before I make the change.

JunTaoLuo replied:
Oh I think I understand what you mean, let me test a few things and I'll make the change if it works.

dsplaisted replied:
Yeah, this is the way the default globs work.  See Microsoft.NET.Sdk.DefaultItems.props.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:
```diff
@@ -1078,6 +1079,12 @@ Copyright (c) .NET Foundation. All rights reserved.
+    Condition="'$(DisableImplicitNamespaceImports)' != 'true'
+      and '$(Language)' == 'C#'
+      and $([MSBuild]::VersionGreaterThanOrEquals('$(_TargetFrameworkVersionWithoutV)', '6.0'))
+      and '$(LangVersion)' == 'preview'"/>
```
We're not going to require the "preview" `LangVersion` for this forever.  Shouldn't the condition be something like this?

```suggestion
      and ('$(LangVersion)' == 'preview' or ($([MSBuild]::VersionGreaterThanOrEquals('$(LangVersion)', '10.0'))))"/>
```

I also wouldn't check the target framework version here.  We should restrict the default `Import` items for C# to .NET 6 and higher, but I think if you're using the appropriate C# version, you should be able to define your own `Import` items and it should work for other target frameworks.

dsplaisted replied:
Also we might want an error or a warning if you define `Import` items but it's not supported by the current language version, otherwise it will be almost impossible to figure out why they aren't working.

pranavkm replied:
Should we remove the LangVersion check entirely? `LangVersion` isn't always a [version string](https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/LanguageVersion.cs#L373-L388) and we shouldn't be parsing compiler code in the SDK. If you remove this, I think the compiler will produce a reasonable error about it. FYI @jaredpar

JunTaoLuo replied:
Looking at this again, I agree that we shouldn't be in the business of resolving langversions (e.g. latest => 10.0) and making decisions based off of it. The compiler error message is pretty clear so I'll remove the lang version check.

I'll also remove the TFM check here.

dsplaisted commented on src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToGenerateImplicitNamespaceImports_DotNet.cs:
```diff
@@ -0,0 +1,196 @@
+        public void It_generates_imports_and_builds_successfully()
+        {
+            var tfm = "net6.0";
+            var testAsset = _testAssetsManager
+                .CopyTestAsset(_testProjectName)
+                .WithSource()
+                .WithTargetFramework(tfm);
```
Here's how to create the project in code via the `TestProject` APIs like I suggested earlier:

```suggestion
            var testProject = new TestProject
            {
                IsExe = true,
                TargetFrameworks = tfm,
            };
            testProject.AdditionalProperties["LangVersion"] = "preview";

            var testAsset = _testAssetsManager.CreateTestProject(testProject);
```

By default the test project name will be set automatically via `CallerMemberNameAttribute`, and you can access it later with the `Name` property in order to construct the expected import filename.

dsplaisted commented on src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToGenerateImplicitNamespaceImports_DotNet.cs:
```diff
@@ -0,0 +1,196 @@
+        }
+
+        [Fact]
+        public void It_can_disable_dotnet_imports()
```
Assuming we move the default `Import` items to a .props file like I suggested, we should also have a test that you can remove specific ones in the project file.

dsplaisted commented on src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToGenerateImplicitNamespaceImports_WebApp.cs:
```diff
@@ -0,0 +1,97 @@
+        public void It_generates_imports_and_builds_successfully()
+        {
+            var tfm = "net6.0";
+            var testAsset = _testAssetsManager
```
When using `TestProject`, you can set the `ProjectSdk` property to `Microsoft.NET.Sdk.Web` to use the Web SDK.

dsplaisted commented on src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.targets:
```diff
@@ -35,6 +36,19 @@ Copyright (c) .NET Foundation. All rights reserved.
     <ProjectCapability Include="AspNetCoreInProcessHosting" />
   </ItemGroup>
 
+  <ItemGroup Condition=" '$(DisableImplicitNamespaceImports_Web)' != 'true'
```
Move this ItemGroup to `Microsoft.NET.Sdk.Web.ProjectSystem.props`

dsplaisted commented:
@wli3 Yes I have context, and I've reviewed this now.

Commit:
Move imports to props

Commit:
Update list of usings

pranavkm commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateImplicitNamespaceImports.targets:
```diff
@@ -0,0 +1,57 @@
+    <ItemGroup>
+      <_UniqueImport Include="@(Import->Distinct())" />
+      <_ImportFileLine Include="// %3Cautogenerated />"/>
+      <_ImportFileLine Include="global using global::%(_UniqueImport.Identity)%3B"/>
```
Should we have the `global::` appear in the itemgroup? This thing would prevent something like

`<Import Include="static Microsoft.AspNetCore.Http.Results" />` from working

davidfowl replied:
No global shouldn't appear in the item group this feature is now aligned with the VB import feature.

Commit:
Remove preview requirements

This works locally, debugging CI

mhutch commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateImplicitNamespaceImports.targets:
```diff
@@ -0,0 +1,57 @@
+    -->
+  <PropertyGroup>
+    <GeneratedImplicitNamespaceImportFile Condition="'$(GeneratedImplicitNamespaceImportFile)' ==''">$(IntermediateOutputPath)$(MSBuildProjectName).ImplicitNamespaceImports$(DefaultLanguageSourceExtension)</GeneratedImplicitNamespaceImportFile>
+    <DisableImplicitNamespaceImports_DotNet Condition="'$(DisableImplicitNamespaceImports_DotNet)' == '' and '$(DisableImplicitNamespaceImports)' != 'true'">false</DisableImplicitNamespaceImports_DotNet>
```
I'm not aware of any common precedent for underscore separators in MSBuild property names. I get the pattern and I don't dislike it, but maybe we should loop in some more folks before establishing new idioms. /cc @rainersigwald @clairernovotny

mhutch replied:
(that said, I'm not suggesting we should block this - we could discuss separately and change it later if necessary)

DamianEdwards replied:
Good point @mhutch. I had coined that on a whim in a chat somewhere and agree that it like deserves more discussion and thought before it sets any kind of precedent. Another case that comes to mind like this are the various default file globbing patterns the various SDKs add today.

@JunTaoLuo as @mhutch says let's not block on this now but it might be something we tweak before GA depending on result of discussions with MSBuild and other folk.

JunTaoLuo replied:
Sounds good

davidfowl replied:
Let's fix the underscoring after this is merged

RussKie commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.DefaultItems.props:
```diff
@@ -82,4 +82,16 @@ Copyright (c) .NET Foundation. All rights reserved.
+    <Import Include="System.Collections.Generic" />
+    <Import Include="System.IO" />
+    <Import Include="System.Linq" />
+    <Import Include="System.Net.Http" />
```
I still maintain `System.Net.Http` shouldn't be in the default set, as it is generally irrelevant in the desktop space.

DamianEdwards replied:
I thought we'd agreed to remove it too. @JunTaoLuo?

mhutch replied:
"Irrelevant" is a bit strong, a lot of desktop/mobile apps use `HttpClient`. That said, their use is probably confined to a few files and may not justify a global namespace.

davidfowl replied:
I think Http should be in the default set.

JunTaoLuo replied:
We discussed this several times already so I'm going to include it for now. The motivation is that we think this is used often enough to warrant being included by default in the base SDK. We had support from Immo on this as well.

Unless this breaks certain scenarios (e.g. type name conflict) I'm not going to remove it here. If you feel this needs more discussion @RussKie, please file an issue to track that discussion.

Commit:
Update conditions

Commit:
Refactor tests and fix blazor wasm

Commit:
Fix build tests

JunTaoLuo commented:
There are still some more tests I'm having trouble with. These tests seem to invoke csc directly with langversion=9.0 which is bizarre. For example:
`Microsoft.NET.Build.Tests.ReferenceExeTests.TestProjectCanReferenceExe(testTemplateName: "xunit") [FAIL]`
```
...
CoreCompile:

C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\MSBuild\Current\Bin\Roslyn\csc.exe /noconfig /unsafe- /checked- /nowarn:1701,1702,1701,1702 /fullpaths /nostdlib+ /platform:x64 /errorreport:prompt /warn:5 
...
/out:obj\Debug\net5.0\win10-x64\ConsoleApp.dll /refout:obj\Debug\net5.0\win10-x64\ref\ConsoleApp.dll 
/target:exe /warnaserror- /utf8output /deterministic+ 

// How is this set?
/langversion:9.0 

...
```

Looks like it's somehow using the VS's csc which uses langversion 9? Not sure why these tests are different from the passing tests which seem to use the csc task in the .NET SDK instead.

Commit:
Restrict LangVersion

JunTaoLuo commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:
```diff
@@ -1078,6 +1079,14 @@ Copyright (c) .NET Foundation. All rights reserved.
+  <Import Project="$(MSBuildThisFileDirectory)Microsoft.NET.GenerateImplicitNamespaceImports.targets"
+    Condition="'$(DisableImplicitNamespaceImports)' != 'true'
+      and '$(Language)' == 'C#'
+      and ('$(LangVersion)' == 'preview'
```
Not a fan of this but until I can figure out why some tests are using VS's csc.exe which specifies LangVersion=9.0, I'm restricting the LangVersion to >= 10.0 workaround the issue.

JunTaoLuo replied:
Anyone know what's the roadmap for C# 10 on Full Framework? I saw some failures on the Full Framework legs where C# 10 isn't available. I'm not sure if that's just because the VS version is out of date or not though.

Commit:
Update

Commit:
Fix type conflict

Commit:
Explicitly set LangVersion in tests

Need to verify who usually sets this

Commit:
Limit subset of tests to core msbuild only

JunTaoLuo commented:
@dsplaisted PTAL. I think I've addressed all outstanding comments and fixed (hopefully all) tests so this is ready for another look.

FYI @davidfowl @DamianEdwards @halter73.

JunTaoLuo commented on src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToGenerateImplicitNamespaceImports_DotNet.cs:
```diff
@@ -0,0 +1,216 @@
+    {
+        public GivenThatWeWantToGenerateImplicitNamespaceImports_DotNet(ITestOutputHelper log) : base(log) { }
+
+        [CoreMSBuildOnlyFact]
```
Restricting some of these tests to Core until I figure out what the C# 10 support story is on Full Framework.

dsplaisted replied:
The problem with full framework is probably that the version of Visual Studio installed on the CI machines, and which is used for the Full Framework tests, is not recent enough.

Instead of using `CoreMSBuildOnlyFact`, try `[RequiresMSBuildVersionFact("17.0.0.32901")]`.  That will make the tests automatically start running on Full Framework once it is updated to a recent enough version.

JunTaoLuo replied:
Ah that's awesome

JunTaoLuo commented on src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToGenerateImplicitNamespaceImports_DotNet.cs:
```diff
@@ -0,0 +1,216 @@
+                TargetFrameworks = tfm,
+                ProjectSdk = "Microsoft.NET.Sdk"
+            };
+            testProject.AdditionalProperties["LangVersion"] = "10.0";
```
Simulates M.CSharp.Core.Targets: https://github.com/dotnet/roslyn/blob/9e30164a1ea97f6f18d48743b05ec68f09fba301/src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets#L20-L22

dsplaisted replied:
If you update stage 0, you shouldn't need this.  You can cherry-pick this commit: a07b803919b8750f461bf7cadf3a7508e64db978

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.DefaultItems.props:
```diff
@@ -82,4 +82,16 @@ Copyright (c) .NET Foundation. All rights reserved.
 
   </ItemGroup>
 
+  <ItemGroup Condition=" '$(DisableImplicitNamespaceImports_DotNet)' != 'true'
```
I think we should move the Visual Basic `Imports` from Microsoft.NET.Sdk.VisualBasic.targets to this file, and deduplicate the ones that are shared.

JunTaoLuo replied:
I was planning on doing that separately to minimize the risk of delaying this PR would that be okay?

dsplaisted replied:
Yes, that would be fine

Commit:
Test changes

Commit:
Remove LangVersion checks

Commit:
Require MSBuild > 17.0.0.32901 for subset of build/publish tests

Commit:
Fix build

Commit:
More full framework MSBuild version requirements

Commit:
Last batch of MSBuild version requirements

Commit:
Fix typo in attribute

Commit:
Skip wpf template test

Commit:
Fix merge conflict with main

Commit:
Rebase fixes

dsplaisted commented:
@sfoslund The Mac OS test leg seems to be timing out, any idea what's going on there?

Commit:
Merge branch 'main' into johluo/global-usings

JunTaoLuo commented:
FYI for debugging purposes, the hang can be observed in https://dev.azure.com/dnceng/public/_build/results?buildId=1226821&view=results. The one that occurred on this PR was lost since I pushed a new commit.

JunTaoLuo commented:
Thanks everyone! I'll file followup issues and announcements listed above.

RussKie commented:
To be honest I'm finding the global usings feature very intrusive. I'm on the 3rd dotnet/* repo that updating to 6.0.100-rc.1.*, and I have to deal with `error CS8652: The feature 'global using directive' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' lang uage version.` errors. 
Even though I participated in the disucssion of the original implementation I still have to go and search GitHub for the property I need to set to false because I don't want to update thousands of source files removing usings. I think an average developer wouldn't know what to do, and will just get massively pissed off.

❗  Btw please note the typo in `lang uage version` text.

davidfowl commented:
A couple of questions:
- What types of projects are these? 
- Are you using VS 2019?

RussKie commented:
dotnet/winforms, dotnet/winforms-designer, dotnet/wpf
Dev17 latest

davidfowl commented:
I don't think those repositories are representative of what the average developer will experience. Have you used the global usings in a real project that doesn't use arcade based repository? Also:

- VS 2019 doesn't fully support C# 10
- Are these projects .NET 6 or .NET standard?

RussKie commented:
The point I'm trying to make - if a developer tries to migrate an existing project to .NET 6.0 right now the developer will be likely be "greeted" with the error above.
I understand the desire to use global usings in new projects, but for existing established projects having it on by default IMO is an unnecessary churn with no real value add.

davidfowl commented:
> The point I'm trying to make - if a developer tries to migrate an existing project to .NET 6.0 right now the developer will be likely be "greeted" with the error above.

Right, specifically because they will be trying to use .NET 6 with VS 2019 where C# 10 support is dodgy.

> I understand the desire to use global usings in new projects, but for existing established projects having it on by default IMO is an unnecessary churn with no real value add.

That's why the option exists to turn it off.

dotMorten commented:
I'm a bit confused here. The issue this PR references only discusses `sdk.web` and `sdk.worker`, but it seems the basic .NET SDK is also addressed in this PR?

In general this feature IMHO causes more issues than it solves. The errors it can introduce in your build are very unhelpful to discover what the root cause is. There is no where in your code or project settings where those global namespaces are defined, since it is all implicit. I'm already [seeing a slew of PRs in the .net repos](https://github.com/search?q=DisableImplicitNamespaceImports&type=issues) opting out of this behavior, which to me indicates the problem with this.

I can also see that WinForms and WPF are planning to add more implicit namespaces via their SDKs, and I wouldn't be surprised if more nuget packages will start doing the same thing. This means that as you're adding dependencies, your code could immediately break, and it will quickly go completely out of hand.

While I do think the idea of global namespaces are great, I'm rather concerned about implicitly defining them. I'd rather this was done at a per-project level instead, either via an `Imports.cs` file, or a project setting that explicitly opted-in instead, and the project templates took advantage of it. I get the goal here is to make your first hello-world project easier to get started with, but I still think that's possible with your first "dotnet new [...]" if it is in the template. I am however worried about the effect this has on every existing project out there, and it affects every single person  who is already over their initial hello-world stage.

lukemcdo commented:
I'd like to point out that the issues aren't just for old code, they're for new code too. Adding a global using to a project template might already be breaking, but removing one from a project template is now a massive consideration that will break third party docs that reference the template.

And if you can't use them in templates for above reason then what's the point? You're still showing new users all of the namespaces anyway then, realistically.

davidfowl commented:
I'm not convinced this is a bad idea yet, I think we have to make a few tweaks to deal with the set of issues we've seen so far but I haven't seen enough to make me want to pull the plug:

- Our repositories aren't a good representation of what users will run into.
- Most people are still on VS 2019 which doesn't fully support C# 10.

 Some common problems that I’ve seen in the wild so far:

- Multi-targeted projects end up complaining about the C# compiler version.
- Projects that extend MSBuild get conflicts between MSBuild Task and System.Threading .Task 
- VS2019 complains about conflicts because it supports a subset of C# 10
- Explicitly specifying the C# version to a version < 10 results in errors and you need to disable the global usings when this happens. 

>  Adding a global using to a project template might already be breaking, but removing one from a project template is now a massive consideration that will break third party docs that reference the template.

Yes that's a known issue. This is why we're being conservative with what we add to them. They version nicely with TFMs so an explicit upgrade is required to see new implicit usings.

I think there are a few tactical things we can do here:

- Disable implicit usings when in VS earlier than 2022 (not sure we can do anything for omnisharp). Or maybe we can detect it some other way
- Disable implicit usings when the language version is explicitly set to < 10

lukemcdo commented:
Fair enough, but like, I still don't understand the benefit. Really feels like the goal here is to smooth the very beginning first lines of C#, but like, this doesn't make the mid-level of learning C# (and to be clear this seems very specific to C#) any more straightforward. 

I also don't like the way that it creates side effects across files. There was a point where the mono project was trying to do single-file projects for example, with the idea being you could just import via copy/paste an entire library. Some projects still aim for this.

Could we get something like "using Globals" or something stupid? I don't know, anything is better than this.

dotMorten commented:
> Most people are still on VS 2019 which doesn't fully support C# 10.
[...]
> Disable implicit usings when the language version is explicitly set to < 10

That's merely a temporary problem.

> Disable implicit usings when in VS earlier than 2022

I'd discourage that. It's nice you can use newer and older VS (depending on user) and still work and compile the same way.

> we're being conservative with what we add to them

Browsing the recent commits in the various .NET repos, 'conservative' isn't what I would call adding more and more globals with each dependency (xamarin, wpf, winforms etc). And with it currently just being an item in an itemgroup, we'll start seeing nuget packages starting to abuse this as well.

davidfowl commented:
> That's merely a temporary problem/solution.

Based on what I've seen that's the bulk of the problem that I've seen so far. With a smattering of ambiguous types (MSBuild Task vs Threading Task)

> I'd discourage that. It's nice you can use newer and older VS (depending on user) and still work and compile the same way.

Maybe, "compile the same way" is impossible when the older version of VS doesn't support the language features.

> Browsing the recent commits in the various .NET repos, 'conservative' isn't what I would call adding more and more globals with each dependency (xamarin, wpf, winforms etc). And with it currently just being an item in an itemgroup, we'll start seeing nuget packages starting to abuse this as well.

We're still in "wait and see" mode.

RussKie commented:
I'm on Dev 17.0.0 Preview 3.0 [31523.460.main], have 6.0.100-rc.1.21375.6 installed globally, trying to build a .NET 6.0 project targeting C# 9.0. 

Why am I getting this error?
![image](https://user-images.githubusercontent.com/4403806/127610379-a1302852-910f-4b38-bcc0-bf6b18872d2e.png)

davidfowl commented:
Because you downgraded the language version to 9 and we currently don't gate the feature on that (though it's being discussed).

swythan commented:
> We're still in "wait and see" mode.

@davidfowl Where is the best place to provide feedback on this? Commenting/voting on a closed PR seems likely to be ineffective.

davidfowl commented:
[Here](https://github.com/dotnet/sdk/issues/18825) is a good place for now.
